### ISSUE TITLE
[feature] Multiple Worker ReplicaSet or Worker Groups

### ISSUE THREAD

**Issue by TaylorHere** [kubeflow/mpi-operator#384]
We have a case that should run Multiple Worker Groups, under which, nodes has different labels and taints, so we are looking forward to the worker spec should a list of ReplicaSet, so that we can have different node select and tolerates.

**Comment by gaocegege** [comment]
Can you please give us an example for the feature?

Maybe like this?

```
worker-normal:
  <podTemplate>
worker-with-different-labels:
  <podTemplate>
```

**Comment by alculquicondor** [comment]
Can you share more about your use case? In which scenarios is it useful to have workers on different types of nodes? Are there other reasons why you would need different templates for workers? Do you have any requirements around the names that the pods would get?

**Comment by TaylorHere** [comment]
Hi，I have an example here:
https://github.com/TaylorHere/mpi-operator/blob/master/examples/v3beta1/example.yaml

we have a reinforcement learning MPI program that should run on different types of nodes, like CPU nodes and GPU nodes

**Comment by gaocegege** [comment]
> Hi，I have an example here:
> https://github.com/TaylorHere/mpi-operator/blob/master/examples/v3beta1/example.yaml
> 
> we have a reinforcement learning MPI program that should run on different types of nodes, like CPU nodes and GPU nodes

Are GPU nodes used for learners, and CPU nodes for actors?

**Comment by zw0610** [comment]
> Hi，I have an example here:
> https://github.com/TaylorHere/mpi-operator/blob/master/examples/v3beta1/example.yaml
> 
> we have a reinforcement learning MPI program that should run on different types of nodes, like CPU nodes and GPU nodes

Will the launcher container launches processes on all the different workers? Or the worker will start the training process by itself?

**Comment by TaylorHere** [comment]
> > Hi，I have an example here:
> > https://github.com/TaylorHere/mpi-operator/blob/master/examples/v3beta1/example.yaml
> > we have a reinforcement learning MPI program that should run on different types of nodes, like CPU nodes and GPU nodes
> 
> Are GPU nodes used for learners, and CPU nodes for actors?

yes, you are right, 
@zw0610  launcher still launches processes on all workers for now, we will try to split it out, like a program runs env and actors and horovod runs learners, it's still on research.
the case I describe above may not work in my company, the Developer's thought that MPI is some kind hard techniques to use, and people with MPI skills is to hard to hire, so this feature request may not worth to implement,  
despite that, I'll still trying to implement this feature on my fork.
in other wise I think more general purpose MPI program can run on Kubernetes in the feature, and MPI-Operator is the best project to achieve that, so maybe we should consider supporting multi worker group in feature

**Comment by alculquicondor** [comment]
It's better to have a proven use case before implementing a feature. Specially this one that changes the API significantly. Otherwise:

- we have to maintain a feature that nobody uses forever
- we complicate the API for the regular user

**Comment by TaylorHere** [comment]
> It's better to have a proven use case before implementing a feature. Specially this one that changes the API significantly. Otherwise:
> 
> * we have to maintain a feature that nobody uses forever
> * we complicate the API for the regular user
you are absolutely right.
maybe I should close this issue, free to re-open, if someone has new idea.

**Comment by gaocegege** [comment]
Maybe we need a new operator for the RL workloads.

**Comment by zw0610** [comment]
> Maybe we need a new operator for the RL workloads.

exactly, which I'm working on for the all-in-one operator. Shall keep everyone updated once the prototype is ready.
